### PR TITLE
Limit the analyzed text for highlighting

### LINKED ISSUE #27517: Limit the number of analyzed tokens during highlighting
Limit the number of analyzed tokens during highlighting.
The maximum number is be 10K, and could be modified through an index setting.
If the limit is exceeded, return a flag that highlights are returned with partial results

This limit does **NOT**  apply if documents used postings, indexed with offsets or term vectors.

relates to #27038
depending on the implementation approach, can be a duplicate of #26515

### PR DESCRIPTION
**Pull request by mayya-sharipova**
- Introduce index level settings to control the max number of character
to be analyzed for highlighting
- Throw an error if analysis is required on a larger text

Closes #27517

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -118,6 +118,17 @@
...
+
+    /**
+     * A setting describing the maximum number of characters that will be analyzed for a highlight request.
+     * This setting is only applicable when highlighting is requested on a text that was indexed without
+     * offsets or term vectors.
+     * The default maximum of 10000 characters is defensive as for highlighting larger texts,
+     * indexing with offsets or term vectors is recommended.
+     */
+    public static final Setting<Integer> MAX_ANALYZED_OFFSET_SETTING =
+        Setting.intSetting("index.analyze.max_analyzed_offset", 10000, 1, Property.Dynamic, Property.IndexScope);
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/IndexSettings.java]: Maybe `index.highlight.max_analyzed_offset` to indicate that it applies to highlighting only ?

File: core/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 2]
```diff
@@ -713,6 +727,13 @@ private void setMaxTokenCount(int maxTokenCount) {
 
     private void setMaxShingleDiff(int maxShingleDiff) { this.maxShingleDiff = maxShingleDiff; }
 
+    /**
+     *  Returns the maximum number of chars that will be analyzed in a highlight request
+     */
+    public int getMaxAnalyzedOffset() { return this.maxAnalyzedOffset; }
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/IndexSettings.java]: Same here, getHighlightMaxAnalyzedOffset ?

**Review by jimczi — CHANGES_REQUESTED** [review]
The changes look good, I left some comment regarding naming.

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by mayya-sharipova** [comment]
@jimczi thanks for the review, Jim

**Comment by Bargs** [comment]
Hi @mayya-sharipova and @jimczi. I was just testing this change and it looks like it's going to break Kibana for users with large fields. It's quite common for users to have fields with >10000 characters so with this change in 6.2 Kibana's Discover (which uses highlighting) will start throwing errors. 

Is it necessary to throw an error? Without the error, users might see a change in highlighting on large fields but at least Kibana wouldn't stop working. The downside is that users may think highlighting is broken if they see matching text that is not highlighted. That might be an acceptable tradeoff and breaking change in a major release but I think either change might be surprising in a minor.

screen shot 2018-01-09 at 10 08 05 am

**Comment by jimczi** [comment]
> That might be an acceptable tradeoff and breaking change in a major release but I think either change might be surprising in a minor

The setting should be unlimited by default in 6.2. @mayya-sharipova can you change the value for that branch and maybe add a deprecation warning for texts bigger than the default limit in 7 ?

**Comment by mayya-sharipova** [comment]
@Bargs thank you for reporting this. Have you tested this on 6.2 and experienced the error? 
Because for 6.2 we don't throw an error, we just issue a deprecation warning: 
https://github.com/elastic/elasticsearch/blob/6.x/core/src/main/java/org/apache/lucene/search/uhighlight/CustomUnifiedHighlighter.java#L133

**Comment by mayya-sharipova** [comment]
@jimczi So,  for 6.2 do we want to do following:
if the settings != LIMIT_& and text exceeds the limit, we will issue an error
if the setting == LIMIT_7  and text exceeds the limit, we will issue a deprecation warning happening in the major version

Sorry, slack has connection problems for me, can't open it for 30 mins

**Comment by Bargs** [comment]
Sorry, like a fool I only tested quickly in master. If 6.2 doesn't respond with an error, Kibana should still work fine. Though we might end up spamming their deprecation logs on every search request.

**Comment by jimczi** [comment]
> So, for 6.2 do we want to do following:
if the settings != LIMIT_7 and text exceeds the limit, we will issue an error
if the setting == LIMIT_7 and text exceeds the limit, we will issue a deprecation warning happening in the major version

For 6.2 I would set the limit to -1 and allow updating this setting only if the value is > 0. This way if the value is -1 you know that the setting is not explicitly set. Then if the value is -1 you can issue a deprecation warning when the text exceeds the default value set in 7 (`10,000`) and if it's not -1 you can throw an error if the text exceeds the provided threshold.

**Comment by kiju98** [comment]
👍

**Pull request merged**